### ISSUE TITLE
Incorrect results are returned if more characters are typed

### ISSUE THREAD

**Issue by WeissP** [armindarvish/consult-omni#36]
Incorrect results are returned if more characters are typed.

**To Reproduce**
1. Run the command `emacs -q --load "debug.el"` where the content of `debug.el` is:
```elisp
(require 'vertico)
(require 'consult-omni)

(setq consult-omni-multi-sources '("Buffer"))
(setq consult-omni-sources-modules-to-load '(consult-omni-buffer))
(require 'consult-omni-sources)
(consult-omni-sources-load-modules)

(find-file "buffer-foo")
(find-file "buffer-bar")

(vertico-mode)
(consult-omni-buffer)
```

2. Type `buf` and wait until results show up. Up to this point, the results are still correct.
3. Type one more `f`; you will see that the results are duplicated:
![flameshot-capture](https://github.com/user-attachments/assets/aaadfe8d-51a2-46cd-a3c0-0406b8216cb8)
4. Remove that `f`, and the results are duplicated again.

**Important Information:**
- OS: Linux
- Version of Emacs: 29.4
- Version of `consult`: 1.8
- Version of `consult-omni`: latest

**Comment by armindarvish** [comment]
@WeissP I cannot reproduce this. See the screenshot below:  

Are you sure you do not have any other setting?  

Can you evaluate the code below after you create the two buffers, and post the results here.  

```emacs-lisp
(let ((query "bufff")
      (buffers (consult--buffer-query :sort 'visibility
                                       :as #'consult--buffer-pair)))
  (mapcar (lambda (item)
          (if (consp item) (setq item (or (car-safe item) item)))
          (when (string-match (concat ".*" query ".*") item)
            (propertize item
                        :title item
                        :url nil
                        :query query
                        :search-url nil)))
        buffers))
```

**Comment by WeissP** [comment]
> Are you sure you do not have any other setting?

Yes, I am sure. The only thing I'm concerned about is that my Emacs is installed via Nix, but I don't really think there will be any differences. 

> Can you evaluate the code below after you create the two buffers.

The results are `(nil nil nil nil)`.

I think the issue is that the old candidate results are never removed. Here is the screenshot if I run Emacs using `emacs -q --load 'debug.el'`:

```elisp
(add-to-list 'load-path "/home/weiss/projects/consult")
(add-to-list 'load-path "/home/weiss/projects/compat/")
(add-to-list 'load-path "/home/weiss/projects/consult-omni")
(add-to-list 'load-path "/home/weiss/projects/consult-omni/sources/")

(require 'compat)
(require 'consult)
(require 'consult-omni)

(find-file "buffer-foo")
(find-file "buffer-bar")

(setq consult-omni-multi-sources '("Buffer"))
(setq consult-omni-sources-modules-to-load '(consult-omni-buffer))
(require 'consult-omni-sources)
(consult-omni-sources-load-modules)

(fido-mode)
(consult-omni-buffer)
```

![output](https://github.com/user-attachments/assets/e7f0fba2-2a61-4270-ad7c-5a62c434b68a)

**Comment by armindarvish** [comment]
@WeissP That's really strange. I cannot reproduce this issue. I just tried with `fido-mode` as well and could not reproduce th behavior in your screenshot.  

Two follow up questions:  

1.  I noticed when you delete the candidates eventually disapper, but that perhaps because the number of chatacters go below 2 and then no results are shown. But, what happens if you delete all the characters and type some other string that should get some hits like "scrat" that should get the scratch buffer? Does the old candidates go away? or if you type more characters that shuould now match a different buffer. For example if you have a buffer called "buffer-test", what happens if you type "buffer-t" do you still see foo and bar as well?

2.  Does `consult-buffer` itself work as expected? how about async or dynamic ones like `consult-grep` and `consult-multi-line`? Do they all work as expected?

**Comment by WeissP** [comment]
> what happens if you delete all the characters and type some other string that should get some hits like "scrat" that should get the scratch buffer?

Once candidates are removed, everything becomes normal, so scratch buffer will show up.

> Does consult-buffer itself work as expected? 

Yes

> How about async or dynamic ones like consult-grep and consult-multi-line? 

They don't work in omni version.
 

By the way, I have tried my MacBook and I can still reproduce this issue. Here are the steps:
1. Install emacs 29.4-1 using https://emacsformacosx.com/ 
2. Git clone Emacs packages `compat`, `consult` and `consult-omni`, using the master/main branch hosted on GitHub.
3. Run `emacs -Q --load 'debug.el'` with the contents I posted above. 

I hope this time you can reproduce this issue, as we should have the same environment.

**Comment by armindarvish** [comment]
@WeissP Ok, I was able to reproduce the issue and it seems that it has something to do with consult HEAD. I could not reproduce this before because I was installing `consult` with `use-package` macro, which installs the latest release (version 1.8) rather than the latest commit at HEAD.  

I have to find some time to see what @minad has done in recent commits in consult and make changes accordingly. At this point I am not sure if this only about changes in `consult-buffer` or more fundamental changes to consult machinery, but I am guessing it is the former and not the latter,  so I should be able to fix it fairly quickly once I find time to look into it.  

In the menawhile see if you can consult 1.8 here: <https://github.com/minad/consult/releases/tag/1.8>  

and let me know if that works.

**Comment by WeissP** [comment]
Yes, the issue seems to be resolved if I'm using the latest release version. Thanks!

A pretty strange issue, to be honest, hope you can sort it out!

**Comment by Gleek** [comment]
Now that 1.9 is released. This issue happens on the current latest version of consult.

**Comment by armindarvish** [comment]
@Gleek I pushed a fix for that. Check the latest version and let me knwo if the issue is resolved.

**Comment by Gleek** [comment]
@armindarvish Works for me now! Thank you.

**Comment by armindarvish** [comment]
Closing this now, as it was fixed in #40!
